### PR TITLE
[filebeat][awss3] - Added support for parquet decoding and decoder config 

### LINKED ISSUE #34662: [Filebeat S3 Input] Add support for Apache Parquet files
With the launch of [Amazon Security Lake](https://aws.amazon.com/security-lake/), users can ingest and store large volumes of data from both AWS services and 3rd party sources (known as providers). SIEM vendors like Elastic can become 'subscribers' by allowing our users to ingest data from the security lake for analysis in Elastic Security. Our goal is to become a subscriber via an integration on our side. The issue to track this integration is here: https://github.com/elastic/integrations/issues/5286 

Security Lake stores the data in S3 buckets within [Parquet formatted files](https://parquet.apache.org/docs/), to avail of the efficient compression with Parquet provides. Filebeat's S3 input does not currently support Parquet files and we need to add support before we can start on the Security Lake integration.

### PR DESCRIPTION
**Pull request by ShourieG**
## Type of change
- Enhancement
- Docs

## What does this PR do?
This PR adds support for a new decoding config option inside the readerConfig struct along with support for 
parquet file decoding using the libbeat parquet reader. The decoding config is created in such a manner that 
in future we will be able to add more decoding codes as well as migrate decoding processes  for JSON, NDJSON 
files which currently occur based on the contentType config option.

An example of the new decoding config:
```yaml
  decoding.codec.parquet.enabled: true
  decoding.codec.parquet.process_parallel: true
  decoding.codec.parquet.batch_size: 1000
``` 

## Why is it important?
This change allows us to officially support parquet decoding for the s3 input and also enable integrations like 
amazon security lake.

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
~~- [ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc`.

## Related issues
- Closes #34662

### CODE REVIEW

**Comment by ShourieG** [comment]
@andrewkroh, not sure why the linux packaging is failing, I could see a kill signal is generated for the following step 
 **go build github.com/aws/aws-sdk-go-v2/service/ec2**  as the compilation fails. But no idea why this is happening.

File: x-pack/filebeat/input/awss3/decoding.go
Comment (belimawr) [inline x-pack/filebeat/input/awss3/decoding.go:32]: [Suggestion]

Return a no-op decoder here, that way the caller of `newDecoder` only needs to check the error, if there is no error, the returned decoder is guaranteed to work.

**Review by belimawr — APPROVED** [review]

**Review by andrewkroh — COMMENTED** [review]
There are a few design questions/decisions I would like to consider.

- If we were designing a decoder interface that we wanted to be able to apply other input sources, what would that API be? Imagine we wanted to be able to decode parquet files in GCS or simply from disk (filestream).

- Should it allow decoder chaining? Why or why not?

- What's relationship (if any) between parser and decoders?

File: x-pack/filebeat/input/awss3/config.go
Comment (ShourieG) [reply x-pack/filebeat/input/awss3/config.go:168]: agree with this, will update this

**Comment by ShourieG** [comment]
@andrewkroh wrt the questions you had - 

1) If we were designing a decoder interface that we wanted to be able to apply other input sources, what would that API be? Imagine we wanted to be able to decode parquet files in GCS or simply from disk (filestream).
 
 - I think if we were to have a generic decoder interface applicable for all inputs, then I would have the decoder API accept a decoder specific config and a 
   io.Reader stream for initialization. For the decode func, it would ideally return a data output channel and an error channel. This would be done so that any 
   input specific logic is kept out of the decoder and the input can easily listen on the channels and extract the data along any errors.

2) Should it allow decoder chaining? Why or why not?
    - I feel chaining would not be a necessity and the decoder config should be extensive enough to define the exact behavior. This also helps keep the api simple
      to use.

3) What's relationship (if any) between parser and decoders?
 - If it's a generic decoder then it should not be concerned with parsing the as and that is the job of the input. Hence the channeled output to 
   to keep any parsing specific logic out of the decoder itself.

For the current implementation, since its a decoder that is specific to the input, these design decisions were not made.

**Comment by ShourieG** [comment]
@andrewkroh I've refactored the decoder to be more generic, updated the tests and docs and added necessary comments. For now I've kept the implementation simple and very basic suited for the current use case but this can be easily extended in future. I've not added a no-op decoder here since I need the nil value returned to branch on the legacy path. I've tried to make as little modifications as possible to the legacy logic so that complications are avoided.

**Comment by ShourieG** [comment]
@andrewkroh apparently the cross-compilation issues are still happening even after the 12.0.1 upgrade as now the error states its requires go 1.20. Need some advice here.

```code
Building using: cmd='build/mage-linux-amd64 golangCrossBuild', env=[CC=oa64-clang, CXX=oa64-clang++, GOARCH=arm64, GOARM=, GOOS=darwin, PLATFORM_ID=darwin-arm64]

[2023-06-13T14:37:09.617Z] # github.com/apache/arrow/go/v12/parquet/internal/utils

[2023-06-13T14:37:09.617Z] /go/pkg/mod/github.com/apache/arrow/go/v12@v12.0.1-0.20230605094802-c153c6d36ccf/parquet/internal/utils/bit_reader.go:230:18: undefined: unpack32

[2023-06-13T14:37:09.618Z] /go/pkg/mod/github.com/apache/arrow/go/v12@v12.0.1-0.20230605094802-c153c6d36ccf/parquet/internal/utils/bit_reader.go:274:3: undefined: BytesToBools

[2023-06-13T14:37:09.618Z] /go/pkg/mod/github.com/apache/arrow/go/v12@v12.0.1-0.20230605094802-c153c6d36ccf/parquet/internal/utils/bit_reader.go:318:18: undefined: unpack32

[2023-06-13T14:37:09.618Z] note: module requires Go 1.20

[2023-06-13T14:40:29.889Z] # github.com/apache/arrow/go/v12/parquet/internal/utils

[2023-06-13T14:40:29.889Z] /go/pkg/mod/github.com/apache/arrow/go/v12@v12.0.1-0.20230605094802-c153c6d36ccf/parquet/internal/utils/bit_reader.go:230:18: undefined: unpack32

[2023-06-13T14:40:29.889Z] /go/pkg/mod/github.com/apache/arrow/go/v12@v12.0.1-0.20230605094802-c153c6d36ccf/parquet/internal/utils/bit_reader.go:274:3: undefined: BytesToBools

[2023-06-13T14:40:29.889Z] /go/pkg/mod/github.com/apache/arrow/go/v12@v12.0.1-0.20230605094802-c153c6d36ccf/parquet/internal/utils/bit_reader.go:318:18: undefined: unpack32

[2023-06-13T14:40:29.889Z] note: module requires Go 1.20

[2023-06-13T14:41:03.168Z] Error: running "go build -o build/golang-crossbuild/filebeat-windows-386.exe -buildmode pie -trimpath -ldflags -s -extldflags=-Wl,--nxcompat -X github.com/elastic/beats/v7/libbeat/version.buildTime=2023-06-13T14:32:06Z -X github.com/elastic/beats/v7/libbeat/version.commit=392e89ec3eed50c1ac607a2d91a1616194abd36a" failed with exit code 2

[2023-06-13T14:41:03.168Z] Error: failed building for windows/386: exit status 2

[2023-06-13T14:41:03.168Z] failed building for windows/386: exit status 2

[2023-06-13T14:43:05.236Z] Error: running "go build -o build/golang-crossbuild/filebeat-linux-386 -buildmode pie -trimpath -tags=withjournald -ldflags -s -X github.com/elastic/beats/v7/libbeat/version.buildTime=2023-06-13T14:33:40Z -X github.com/elastic/beats/v7/libbeat/version.commit=392e89ec3eed50c1ac607a2d91a1616194abd36a" failed with exit code 2

[2023-06-13T14:43:05.236Z] Error: failed building for linux/386: exit status 2

[2023-06-13T14:43:05.236Z] failed building for linux/386: exit status 2
```

**Comment by ShourieG** [comment]
@andrewkroh could you review the current implementation while we wait for the CI pipeline to be fixed, so that we can merge before FF, that would be really great as Crest will be taking up the security lake integration after this merge. 

I opened a public issue: https://github.com/apache/arrow/issues/36052 for the cross build errors on 32bit systems and they have been resolved with a recent PR, but still not updating the library since there is no stable release out with these fixes.

**Comment by faec** [comment]
Build fix is in review: https://github.com/elastic/beats/pull/35789

File: x-pack/filebeat/input/awss3/s3_objects.go
Comment (andrewkroh) [inline x-pack/filebeat/input/awss3/s3_objects.go:281]: If this JSON array contains more than one event then they all will have the same offset. The `offset` is meant to hold the byte index of the first character of the JSON object.

If two events have the same offset then they have the same `_id` which means one of them will be discarded as a duplicate.

Comment (ShourieG) [reply x-pack/filebeat/input/awss3/s3_objects.go]: @andrewkroh agree with this and have updated the logic to track offsets more effectively. I have concerns that this might not solve the issue in some cases. Since the parquet reader does not return a traditional offset and the io.Reader is created with every fresh byte slice returned, every slice batch that it returns will ideally start with the same offset when passed to the JSON decoder. i.e If there are 3000 records with each row and JSON object having the same size where for example batch size is 1000, then the 3 batches would have the same offset for all the objects even though each record in a 1000 batch have different offsets to each other. That would mean it would publish the first 1000 records and discard the rest 2000 ?

Comment (andrewkroh) [reply x-pack/filebeat/input/awss3/s3_objects.go]: Are events always returned in the same order from the parquet reader? If so, then perhaps instead of treating the offset as a byte index we use it as an event index.

Comment (ShourieG) [reply x-pack/filebeat/input/awss3/s3_objects.go]: Yes the parquet reader always maintains the order of events .

Comment (ShourieG) [reply x-pack/filebeat/input/awss3/s3_objects.go]: Treating the offset as an event index should work. Would anything need to change here or does it look good for merging for now ?

File: x-pack/filebeat/input/awss3/decoding.go
Comment (ShourieG) [reply x-pack/filebeat/input/awss3/decoding.go:32]: Right now we require the nil value to branch on legacy paths, so will do this once all the legacy paths are removed.

File: x-pack/filebeat/docs/inputs/input-aws-s3.asciidoc
Comment (zmoog) [inline x-pack/filebeat/docs/inputs/input-aws-s3.asciidoc:153]: I was surprised to find out that we can use config this way (`decoding.codec.parquet: enabled`).

Since we define the struct as:

```golang
type parquetCodecConfig struct {
	Enabled         bool `config:"enabled"`
	ProcessParallel bool `config:"process_parallel"`
	BatchSize       int  `config:"batch_size" default:"1"`
}
```

I would expect something along the line of:

```yaml
decoding.codec.parquet.enabled = true
...
```

Good to know.

**Review by zmoog — APPROVED** [review]

File: x-pack/filebeat/docs/inputs/input-aws-s3.asciidoc
Comment (ShourieG) [reply x-pack/filebeat/docs/inputs/input-aws-s3.asciidoc:153]: actually all of them should ideally be ":" and not "=" since the config is an yaml, thanks for the review, will need to update the docs.

Comment (zmoog) [reply x-pack/filebeat/docs/inputs/input-aws-s3.asciidoc:153]: Yeah, I mindlessly followed the example below, my bad.

However, I was surprised by `decoding.codec.parquet: enabled` vs. `decoding.codec.parquet.enabled: true`.

Thanks for confirming.

Comment (ShourieG) [reply x-pack/filebeat/docs/inputs/input-aws-s3.asciidoc]: yea it should be decoding.codec.parquet.enabled: true not decoding.codec.parquet: enabled, it was a typo in the docs, updated it now.

File: x-pack/filebeat/input/awss3/s3_objects.go
Comment (ShourieG) [reply x-pack/filebeat/input/awss3/s3_objects.go]: added event offset tracking

**Comment by yago82** [comment]
Hello everyone,

I'm facing an issue while trying to retrieve a Parquet file from S3 using Filebeat. Below, I've included configuration details:

```
filebeat.inputs:
- type: aws-s3
  bucket_arn: ${BUCKET_ARN}
  bucket_list_prefix: ${BUCKET_LIST_PREFIX}
  bucket_list_interval: 60s
  region: eu-west-1
  default_region: eu-west-1
  number_of_workers: 5
  access_key_id: ${ACCESS_KEY_ID}
  secret_access_key: ${SECRET_ACCESS_KEY}
  decoding.codec.parquet.enabled: true
  decoding.codec.parquet.process_parallel: true
  decoding.codec.parquet.batch_size: 1000

setup.template.enabled: false

processors:
  - add_fields:
      target: '@metadata'
      fields:
        op_type: "index"

output.elasticsearch:
  hosts: ["${ELASTICSEARCH_HOSTS}"]
  username: ${ELASTICSEARCH_USERNAME}
  password: ${ELASTICSEARCH_PASSWORD}
  protocol: https
  index: utenti123
  allow_older_versions: true
```

I have tried various bucket_list_prefix solutions including:

```
emr-serverless/user-output/
emr-serverless/user--output//
emr-serverless/user-output/*
emr-serverless/user-output/*/
```

However, we consistently encounter the following error:

`failed processing S3 event for object key "emr-serverless/user-output/" in bucket "root-content": failed to create parquet decoder: failed to create parquet reader: parquet: file too small (size=0)`

Any insights or suggestions on troubleshooting steps would be highly appreciated. Please let me know if additional information is needed.

Thank you

**Comment by andrewkroh** [comment]
Trying using `file_selectors` to selectively apply the parquet decoding to specific files.

https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-aws-s3.html#_file_selectors

**Comment by kdHub** [comment]
Thanks for the suggestion on file_selectors. I was using them and did not realize my issue until seeing this last comment and moving the decoding block into file_selector section solved my problem.

**Pull request merged**
